aot-std-optimizers/src/main/java/io/micronaut/aot/std/sourcegen/DeduceEnvironmentSourceGenerator.java

```diff
@@ -0,0 +1,99 @@
+import io.micronaut.aot.core.codegen.AbstractCodeGenerator;
+import io.micronaut.aot.core.context.ApplicationContextAnalyzer;
+import io.micronaut.context.ApplicationContextBuilder;
+import io.micronaut.context.ApplicationContextConfiguration;
+import io.micronaut.context.ApplicationContextConfigurer;
+import io.micronaut.context.BeanContextConfiguration;
+
+import java.util.Collection;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static javax.lang.model.element.Modifier.PUBLIC;
+
+/**
+ * This code generator is responsible for taking the result
+ * of the environment deduction, which is the set of active
+ * environment names and the package names, and inject is
+ * via a custom application context configurer. The resulting
+ * class will effectively disable environment deduction, so
+ * it will be done at build time instead of run time.
+ */
+@AOTModule(
+        id = DeduceEnvironmentSourceGenerator.ID,
+        description = DeduceEnvironmentSourceGenerator.DESCRIPTION
+)
+public class DeduceEnvironmentSourceGenerator extends AbstractCodeGenerator {
+    public static final String ID = "deduce.environment";
+    public static final String DESCRIPTION = "Deduces the environment at build time instead of runtime";
+    public static final String DEDUCED_ENVIRONMENT_CONFIGURER = "DeducedEnvironmentConfigurer";
+
+    @Override
+    public void generate(AOTContext context) {
+        ApplicationContextAnalyzer analyzer = context.getAnalyzer();
+        Set<String> environmentNames = analyzer.getEnvironmentNames();
+        BeanContextConfiguration contextConfiguration = analyzer.getApplicationContext().getContextConfiguration();
+        if (contextConfiguration instanceof ApplicationContextConfiguration) {
+            ((ApplicationContextConfiguration) contextConfiguration).getDeduceEnvironments().ifPresent(deduceEnvironments -> {
+                if (deduceEnvironments) {
+                    Collection<String> packages = analyzer.getApplicationContext().getEnvironment().getPackages();
+                    context.registerGeneratedSourceFile(
+                            context.javaFile(buildApplicationContextConfigurer(environmentNames, packages))
+                    );
+                    writeServiceFile(context, ApplicationContextConfigurer.class, DEDUCED_ENVIRONMENT_CONFIGURER);
+                }
+            });
+        }
+    }
+
+    private TypeSpec buildApplicationContextConfigurer(Set<String> environmentNames, Collection<String> packages) {
+        MethodSpec.Builder bodyBuilder = MethodSpec.methodBuilder("configure")
+                .addModifiers(PUBLIC)
+                .addAnnotation(Override.class)
+                .addParameter(ApplicationContextBuilder.class, "builder");
+        bodyBuilder.addStatement("builder.deduceEnvironment(false)");
+        if (!environmentNames.isEmpty()) {
+            bodyBuilder.addStatement("builder.defaultEnvironments($L)", toQuotedStringList(environmentNames));
```

Comment:
I think this should call `environments` instead of `defaultEnvironment` 🤔

the latter means if the user specifies another environment it will be overridden

Reply:
Right, I'll change it.